### PR TITLE
Am 335 Fully-validate connection parameters on save

### PR DESCRIPTION
**Pull request by ErnestoLoma**

### CODE REVIEW

**Review by cmgrote — CHANGES_REQUESTED** [review]
One other general question — are we now validating responses as well as requests? I don't think we should validate the former, or we may end up not giving back whatever the server tells us and instead giving the user an error (which I don't think is desirable).

Even if an assigned user or admin has been removed, or previously set to some inaccurate value by some direct API call, we should still return the object to a user. We can validate it if they submit those details to change the object — but fundamentally we should only validate this information when sent in a request, not received in a response.

File: pyatlan/cache/group_cache.py [hunk 1]
```diff
@@ -107,3 +116,15 @@ def _get_name_for_id(self, idstr: str) -> Optional[str]:
...
+
+    def _validate_names(self, names: Iterable[str]):
+        """
+        Validate that the given human-readable group names are valid. A ValueError will be raised in any are not.
+
+        :param names: a collection of group names to be checked
+        """
+        for group_alias in names:
+            if not self.get_id_for_alias(group_alias):
+                raise ValueError(
```
[the remark this answers is not in the document]
Comment (ErnestoLoma) [reply pyatlan/cache/group_cache.py:128]: Yes it needs to be a ValueError

File: pyatlan/cache/role_cache.py [hunk 2]
```diff
@@ -86,3 +95,13 @@ def _get_name_for_id(self, idstr: str) -> Optional[str]:
...
+
+    def _validate_idstrs(self, idstrs: Iterable[str]):
+        """
+        Validate that the given role GUIDs are valid. A ValueError will be raised in any are not.
+
+        :param idstrs: a collection of unique identifiers (GUID) of the roles to be checked
+        """
+        for role_id in idstrs:
+            if not self.get_name_for_id(role_id):
+                raise ValueError(f"Provided role ID {role_id} was not found in Atlan.")
```
[the remark this answers is not in the document]
Comment (ErnestoLoma) [reply pyatlan/cache/role_cache.py:107]: Yes

**Comment by ErnestoLoma** [comment]
Validation happens whenever the the object is instantiated. There is is a method that allows an object to be constructed without any validation, but it's all or nothing. If we turn off validation when marshalling objects from JSON I think we run the risk of hiding errors in the code the might have been exposed if an error was raised when handling data that is actually valid.

**Comment by cmgrote** [comment]
> Validation happens whenever the the object is instantiated. There is is a method that allows an object to be constructed without any validation, but it's all or nothing. If we turn off validation when marshalling objects from JSON I think we run the risk of hiding errors in the code the might have been exposed if an error was raised when handling data that is actually valid.

Yeah, I don't want to switch of validation entirely. I'm more thinking there are levels of validation:

- Is the JSON well-formed (always on)
- Is the minimal information in the JSON for us to parse it, e.g. sufficient info to tell us whether it's an asset vs some other object (always on)

Contrasted with:

- Is the information in the JSON *correct* (only for requests a user creates and wants to POST, not used for anything we GET back)

This is akin to the idea of an attribute being "required" or not — answering this question really depends on the context:

- Very few attributes are required in a response (for assets, only a `typeName` and `guid`)
- Other attributes are required for update requests (i.e. `qualifiedName` and `name`)
- Still more attributes are required for requests to create an asset (i.e. `qualifiedName`, `name`, and `qualifiedName` or `guid` (+ `typeName`) of the asset in which the asset-to-create will be contained)

^^ Just like there isn't a simple all-or-nothing "required" for attributes, I don't think we should have a simple all-or-nothing "validated" for payloads — we need to vary the validation based on the context (in particular request vs response).

If Pydantic can't handle the contextual validation, then we should only implement the minimal "always-on" validation via Pydantic, and move the contextual validation out into other methods (e.g. just before invoking `_call_api`, in the worst case).

**Review by cmgrote — APPROVED** [review]

**Pull request merged**